File: design-pattern/02-adapter-pattern/adapterPattern.dart
abstract class Target {
  // ターゲットインタフェース(抽象)
  void request();
}

class Adaptee {
  // 既存クラス(具象)
  void specificRequest() {
    print('既存クラスのメソッドが呼ばれました。');
  }
}

class Adapter implements Target {
  // アダプタークラス(具象)
  // Targetインタフェースを実装する
  late Adaptee _adaptee;

  Adapter(Adaptee adaptee) {
    _adaptee = adaptee;
  }

  @override
  void request() {
    _adaptee.specificRequest();
  }
}

void main() {
  var adaptee = Adaptee();
  var adapter = Adapter(adaptee);
  adapter.request();
}

File: design-pattern/02-adapter-pattern/adapterPattern.ts
interface DirectCurrent {
  // 直流12V
  request(): void;
}

class AlternatingCurrent {
  // 交流100V
  specificRequest(): void {
    console.log("コンセントから交流100Vを提供する。");
  }
}
class ACAdapter implements DirectCurrent {
  // ACアダプター
  private adaptee: AlternatingCurrent;

  constructor(adaptee: AlternatingCurrent) {
    this.adaptee = adaptee;
  }

  request(): void {
    this.adaptee.specificRequest();
    console.log('交流100Vを直流12Vに変換する。')
  }
}

const adaptee = new AlternatingCurrent();
const adapter = new ACAdapter(adaptee);
adapter.request();
